src/samples/plugins/ModularSessionKeyPlugin.sol

```diff
@@ -0,0 +1,336 @@
+            revert WrongDataLength();
+        }
+        for (uint256 i = 0; i < tempOwners.length; i++) {
+            _removeSessionKey(msg.sender, tempOwners[i], allowedSelectors[i]);
+        }
+        emit SessionKeysRemoved(msg.sender, tempOwners, allowedSelectors);
+    }
+
+    // ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    // ┃    Plugin view functions    ┃
+    // ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+
+    /// @inheritdoc IModularSessionKeyPlugin
+    function getSessionDuration(address account, address tempOwner, bytes4 allowedSelector)
+        external
+        view
+        returns (uint48 _validAfter, uint48 _validUntil)
+    {
+        bytes memory key = account.allocateAssociatedStorageKey(0, 1);
+        StoragePointer ptr = key.associatedStorageLookup(keccak256(abi.encodePacked(tempOwner, allowedSelector)));
+        SessionInfo storage sessionInfo = _castPtrToStruct(ptr);
+        _validAfter = sessionInfo.validAfter;
+        _validUntil = sessionInfo.validUntil;
+    }
+
+    // ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    // ┃    Plugin interface functions    ┃
+    // ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+
+    /// @inheritdoc BasePlugin
+    function onInstall(bytes calldata data) external override {
+        if (data.length != 0) {
+            (
+                address[] memory tempOwners,
+                bytes4[] memory allowedSelectors,
+                uint48[] memory validAfters,
+                uint48[] memory validUntils
+            ) = abi.decode(data, (address[], bytes4[], uint48[], uint48[]));
+            if (
+                tempOwners.length != allowedSelectors.length || tempOwners.length != validAfters.length
+                    || tempOwners.length != validUntils.length
+            ) {
+                revert WrongDataLength();
+            }
+            for (uint256 i = 0; i < tempOwners.length; i++) {
+                _addSessionKey(msg.sender, tempOwners[i], allowedSelectors[i], validAfters[i], validUntils[i]);
+            }
+        }
+    }
+
+    /// @inheritdoc BasePlugin
+    function onUninstall(bytes calldata data) external override {
+        if (data.length != 0) {
+            (address[] memory tempOwners, bytes4[] memory allowedSelectors) =
+                abi.decode(data, (address[], bytes4[]));
+            if (tempOwners.length != allowedSelectors.length) {
+                revert WrongDataLength();
+            }
+            for (uint256 i = 0; i < tempOwners.length; i++) {
+                _removeSessionKey(msg.sender, tempOwners[i], allowedSelectors[i]);
+            }
+        }
+    }
+
+    /// @inheritdoc BasePlugin
+    function userOpValidationFunction(uint8 functionId, UserOperation calldata userOp, bytes32 userOpHash)
+        external
+        view
+        override
+        returns (uint256)
+    {
+        if (functionId == uint8(FunctionId.USER_OP_VALIDATION_TEMPORARY_OWNER)) {
+            (address signer,) = userOpHash.toEthSignedMessageHash().tryRecover(userOp.signature);
```

Comment:
Check the result of tryRecover, and if `err != ECDSA.RecoverError.NoError`, revert with `InvalidSignature()`.